module/VuFind/src/VuFind/Controller/NotificationsController.php

```diff
@@ -0,0 +1,433 @@
+<?php
+
+/**
+ * Notifications Controller
+ *
+ * PHP version 8
+ *
+ * Copyright (C) effective WEBWORK GmbH 2023.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.    See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  Controller
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Johannes Schultze <schultze@effective-webwork.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+
+namespace VuFind\Controller;
+
+use ElGigi\CommonMarkEmoji\EmojiExtension;
+use Laminas\ServiceManager\ServiceLocatorInterface;
+use Laminas\Stdlib\ArrayObject;
+use League\CommonMark\Environment\Environment;
+use League\CommonMark\Extension\Autolink\AutolinkExtension;
+use League\CommonMark\Extension\CommonMark\CommonMarkCoreExtension;
+use League\CommonMark\MarkdownConverter;
+use VuFind\Form\BroadcastsForm;
+use VuFind\Form\PagesForm;
+
+/**
+ * Controls the configuration and display of notifications
+ *
+ * @category VuFind
+ * @package  Controller
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Johannes Schultze <schultze@effective-webwork.de>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+class NotificationsController extends \VuFind\Controller\AbstractBase
+{
+    /**
+     * Notifications config
+     *
+     * @var array
+     */
+    private $config;
+
+    /**
+     * Constructor
+     *
+     * @param ServiceLocatorInterface $sm Service locator
+     *
+     * @throws \Psr\Container\ContainerExceptionInterface
+     * @throws \Psr\Container\NotFoundExceptionInterface
+     */
+    public function __construct(ServiceLocatorInterface $sm)
+    {
+        parent::__construct($sm);
+        $this->config = $sm->get(\VuFind\Config\YamlReader::class)->get('Notifications.yaml');
+        $this->accessPermission = 'notifications.Admin';
```

Comment:
This permission requirement is likely the reason that users can't view notifications while logged out. I don't think you want to restrict the read portions of the interface, only the administrative ones... but putting this setting here requires the permission globally for all actions of this controller. I think you have two options:

Option 1: Create a separate ManageNotificationsController for the management parts, and make NotificationsController responsible only for displaying notifications. (Or some similar scheme).

Option 2: Remove the global accessPermission property, and instead perform a permission check at the top of each action that needs to be restricted. You can do this with the [Permission controller plugin](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/Controller/Plugin/Permission.php). You can see an example of this in action in the [AbstractRecord controller](https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFind/Controller/AbstractRecord.php#L457-L461).

Reply:
I just tried option 2, but I don't understand the behaviour. I added this block to the top of `pagesAction()` in the `NotificationsController`

```
        $isAdmin = $this->permission()->check('notifications.Admin', false);
        if (!$this->getAuthManager()->isLoggedIn() || !$isAdmin) {
            return $this->redirect()->toRoute('myresearch-home');
        }
```

`$isAdmin` returns as `null`, which I don't understand. In the `check()`-method of `Permission` this results in `false`:

```
    if ($this->permissionManager->isAuthorized($permission) !== true) {
```

but if I add a temporary variable right above this line, it results in `true`:

```
$testPermission = $this->permissionManager->isAuthorized($permission);
```

When checking the results of `$this->permission()->check()` on other Admin pages, the results for `access.AdminModule` are also `null`, but I can access those pages.

Reply:
@jschultze, the second argument to the check method determines how the check behaves. If you set it to false, as in your example, then the check does nothing and always returns null. The check() method is meant to be used in contexts where you may conditionally want to force login or throw an exception when permissions are missing. If you just want to check whether the current user has a permission, you should use isAuthorized() instead. Does switching your logic to use isAuthorized() instead of check() lead to a situation that makes more sense?

Reply:
@demiankatz Thanks, `isAuthorized()` works and makes much more sense, yes.